cpmpy/solvers/exact.py

```diff
@@ -0,0 +1,528 @@
+        if not self.supported():
+            raise Exception("Install 'exact' as a Python package to use this solver interface")
+
+        from exact import Exact as xct
+
+        assert subsolver is None, "Exact does not allow subsolvers."
+
+        # initialise the native solver object
+        self.xct_solver = xct()
+        self.xct_solver.setOption("inp-purelits", "0")    # no dominance breaking to preserve solutions
+        self.xct_solver.setOption("inp-dombreaklim", "0") # no dominance breaking to preserve solutions
+
+        # for solving with assumption variables,
+        self.assumption_dict = None
+
+        # objective can only be set once, so keep track of this
+        self.solver_is_initialized = False
+        self.objective_given = False
+        self.objective_minimize = True
+
+        # initialise everything else and post the constraints/objective
+        super().__init__(name="exact", cpm_model=cpm_model)
+
+    def _fillObjAndVars(self):
+        if not self.xct_solver.hasSolution():
+            self.objective_value_ = None
+            return False
+
+        # fill in variable values
+        lst_vars = list(self.user_vars)
+        exact_vals = self.xct_solver.getLastSolutionFor(self.solver_vars(lst_vars))
+        for cpm_var, val in zip(lst_vars,exact_vals):
+            cpm_var._value = bool(val) if isinstance(cpm_var, _BoolVarImpl) else val # xct value is always an int
+
+        # translate objective
+        self.objective_value_ = self.xct_solver.getObjectiveBounds()[1] # last upper bound to the objective
+        if not self.objective_minimize:
+            self.objective_value_ = -self.objective_value_
+
+        return True
+
+    def solve(self, time_limit=None, assumptions=None, **kwargs):
+        """
+            Call Exact
+
+            Arguments:
+            - time_limit:  maximum solve time in seconds (float, optional)
+            - assumptions: list of CPMpy Boolean variables (or their negation) that are assumed to be true.
+                           For repeated solving, and/or for use with s.get_core(): if the model is UNSAT,
+                           get_core() returns a small subset of assumption variables that are unsat together.
+
+            Additional keyword arguments:
+            The Exact solver parameters are defined by https://gitlab.com/JoD/exact/-/blob/main/src/Options.hpp#L207
+        """
+        from exact import Exact as xct
+
+        if not self.solver_is_initialized:
+            assert not self.objective_given
+            self.xct_solver.init([],[])
+            self.solver_is_initialized=True
+            self.xct_solver.setOption("verbosity","0")
+        self.xct_solver.clearAssumptions()
+
+        # set additional keyword arguments
+        for (kw, val) in kwargs.items():
+            self.xct_solver.setOption(kw,str(val))
+
+        # set assumptions
+        if assumptions is not None:
+            assert(all(v.is_bool() for v in assumptions))
```

Comment:
This will fail if someone gives a non-CPMpy-expression as assumption, maybe add a clearer error message?

Reply:
Added an error message, is this what you aimed for?